File: backend/handlers/marcas.go
// Package handlers provides HTTP request handlers
package handlers

import (
	"net/http"
	"strconv"

	"github.com/gorilla/mux"
	"github.com/mcmarin9/dolls/database"
	"github.com/mcmarin9/dolls/models"
	"github.com/mcmarin9/dolls/utils"
)

// GetMarcas returns all marcas
func GetMarcas(w http.ResponseWriter, r *http.Request) {
	query := "SELECT id, nombre, created_at FROM marca ORDER BY nombre ASC"

	rows, err := database.ExecuteQuery(query)
	if err != nil {
		respondWithError(w, http.StatusInternalServerError, "Error al obtener marcas")
		return
	}
	defer rows.Close()

	var marcas []models.Marca
	for rows.Next() {
		var marca models.Marca
		if err := rows.Scan(&marca.ID, &marca.Nombre, &marca.CreatedAt); err == nil {
			// Get associated fabricantes for this marca
			fabricantesQuery := `
				SELECT f.id, f.nombre 
				FROM fabricantes f
				INNER JOIN marca_fabricante mf ON f.id = mf.fabricante_id
				WHERE mf.marca_id = ?
				ORDER BY f.nombre ASC
			`
			fabricantesRows, err := database.ExecuteQuery(fabricantesQuery, marca.ID)
			if err == nil {
				defer fabricantesRows.Close()
				var fabricantes []models.Fabricante
				for fabricantesRows.Next() {
					var fabricante models.Fabricante
					if err := fabricantesRows.Scan(&fabricante.ID, &fabricante.Nombre); err == nil {
						fabricantes = append(fabricantes, fabricante)
					}
				}
				marca.Fabricantes = fabricantes
			}
			marcas = append(marcas, marca)
		}
	}

	respondWithJSON(w, http.StatusOK, marcas)
}

// AddMarca creates a new marca
func AddMarca(w http.ResponseWriter, r *http.Request) {
	type MarcaInput struct {
		Nombre       string `json:"nombre"`
		FabricanteID int    `json:"fabricante_id"`
	}

	var input MarcaInput
	if err := parseJSONBody(r, &input); err != nil {
		respondWithError(w, http.StatusBadRequest, "Datos inválidos")
		return
	}

	// Validate input
	if err := utils.ValidateMarcaInput(input.Nombre); err != nil {
		respondWithError(w, http.StatusBadRequest, err.Error())
		return
	}

	if input.FabricanteID <= 0 {
		respondWithError(w, http.StatusBadRequest, "fabricante_id inválido")
		return
	}

	// Validate that fabricante exists
	var fabricanteExists int
	err := database.ExecuteQueryRow("SELECT COUNT(*) FROM fabricantes WHERE id = ?", input.FabricanteID).Scan(&fabricanteExists)
	if err != nil || fabricanteExists == 0 {
		respondWithError(w, http.StatusBadRequest, "Fabricante no encontrado")
		return
	}

	// Insert marca
	insertQuery := "INSERT INTO marca (nombre) VALUES (?)"
	result, err := database.ExecuteUpdate(insertQuery, input.Nombre)
	if err != nil {
		respondWithError(w, http.StatusInternalServerError, "Error al crear marca")
		return
	}

	marcaID, _ := result.LastInsertId()

	// Associate with fabricante
	_, err = database.ExecuteUpdate("INSERT INTO marca_fabricante (marca_id, fabricante_id) VALUES (?, ?)",
		marcaID, input.FabricanteID)
	if err != nil {
		logger.Error("Error creating marca_fabricante relationship", err)
		respondWithError(w, http.StatusInternalServerError, "Error al asociar marca con fabricante")
		return
	}

	respondWithJSON(w, http.StatusCreated, models.SuccessResponse{
		Message: "Marca creada exitosamente",
		Data:    map[string]int64{"id": marcaID},
	})
}

// UpdateMarca updates an existing marca
func UpdateMarca(w http.ResponseWriter, r *http.Request) {
	vars := mux.Vars(r)
	id, err := strconv.Atoi(vars["id"])
	if err != nil {
		respondWithError(w, http.StatusBadRequest, "ID inválido")
		return
	}

	type MarcaInput struct {
		Nombre string `json:"nombre"`
	}

	var input MarcaInput
	if err := parseJSONBody(r, &input); err != nil {
		respondWithError(w, http.StatusBadRequest, "Datos inválidos")
		return
	}

	// Validate input
	if err := utils.ValidateMarcaInput(input.Nombre); err != nil {
		respondWithError(w, http.StatusBadRequest, err.Error())
		return
	}

	// Update marca
	updateQuery := "UPDATE marca SET nombre = ? WHERE id = ?"
	_, err = database.ExecuteUpdate(updateQuery, input.Nombre, id)
	if err != nil {
		respondWithError(w, http.StatusInternalServerError, "Error al actualizar marca")
		return
	}

	respondWithJSON(w, http.StatusOK, models.SuccessResponse{
		Message: "Marca actualizada exitosamente",
	})
}

// DeleteMarca deletes a marca
func DeleteMarca(w http.ResponseWriter, r *http.Request) {
	vars := mux.Vars(r)
	id, err := strconv.Atoi(vars["id"])
	if err != nil {
		respondWithError(w, http.StatusBadRequest, "ID inválido")
		return
	}

	// Check if marca is being used by dolls
	var count int
	database.ExecuteQueryRow("SELECT COUNT(*) FROM dolls WHERE marca_id = ?", id).Scan(&count)
	if count > 0 {
		respondWithError(w, http.StatusBadRequest, "No se puede eliminar: marca en uso por muñecas")
		return
	}

	// Delete from marca_fabricante first
	database.ExecuteUpdate("DELETE FROM marca_fabricante WHERE marca_id = ?", id)

	// Delete marca
	result, err := database.ExecuteUpdate("DELETE FROM marca WHERE id = ?", id)
	if err != nil {
		respondWithError(w, http.StatusInternalServerError, "Error al eliminar marca")
		return
	}

	rowsAffected, _ := result.RowsAffected()
	if rowsAffected == 0 {
		respondWithError(w, http.StatusNotFound, "Marca no encontrada")
		return
	}

	respondWithJSON(w, http.StatusOK, models.SuccessResponse{
		Message: "Marca eliminada exitosamente",
	})
}

// GetFabricantes returns all fabricantes
func GetFabricantes(w http.ResponseWriter, r *http.Request) {
	query := "SELECT id, nombre, created_at FROM fabricantes ORDER BY nombre ASC"

	rows, err := database.ExecuteQuery(query)
	if err != nil {
		respondWithError(w, http.StatusInternalServerError, "Error al obtener fabricantes")
		return
	}
	defer rows.Close()

	var fabricantes []models.Fabricante
	for rows.Next() {
		var fabricante models.Fabricante
		if err := rows.Scan(&fabricante.ID, &fabricante.Nombre, &fabricante.CreatedAt); err == nil {
			fabricantes = append(fabricantes, fabricante)
		}
	}

	respondWithJSON(w, http.StatusOK, fabricantes)
}

// AddFabricante creates a new fabricante
func AddFabricante(w http.ResponseWriter, r *http.Request) {
	type FabricanteInput struct {
		Nombre string `json:"nombre"`
	}

	var input FabricanteInput
	if err := parseJSONBody(r, &input); err != nil {
		respondWithError(w, http.StatusBadRequest, "Datos inválidos")
		return
	}

	// Validate input
	if err := utils.ValidateMarcaInput(input.Nombre); err != nil {
		respondWithError(w, http.StatusBadRequest, err.Error())
		return
	}

	// Check if fabricante already exists
	var count int
	err := database.ExecuteQueryRow("SELECT COUNT(*) FROM fabricantes WHERE nombre = ?", input.Nombre).Scan(&count)
	if err == nil && count > 0 {
		respondWithError(w, http.StatusBadRequest, "Este fabricante ya existe")
		return
	}

	// Insert fabricante
	insertQuery := "INSERT INTO fabricantes (nombre) VALUES (?)"
	result, err := database.ExecuteUpdate(insertQuery, input.Nombre)
	if err != nil {
		respondWithError(w, http.StatusInternalServerError, "Error al crear fabricante")
		return
	}

	fabricanteID, _ := result.LastInsertId()

	// Fetch the created fabricante
	var fabricante models.Fabricante
	err = database.ExecuteQueryRow("SELECT id, nombre, created_at FROM fabricantes WHERE id = ?", fabricanteID).
		Scan(&fabricante.ID, &fabricante.Nombre, &fabricante.CreatedAt)
	if err != nil {
		respondWithError(w, http.StatusInternalServerError, "Error al recuperar fabricante creado")
		return
	}

	respondWithJSON(w, http.StatusCreated, fabricante)
}

// GetMarcasByFabricante returns all marcas for a specific fabricante
func GetMarcasByFabricante(w http.ResponseWriter, r *http.Request) {
	vars := mux.Vars(r)
	fabricanteID, err := strconv.Atoi(vars["fabricante_id"])
	if err != nil {
		respondWithError(w, http.StatusBadRequest, "fabricante_id inválido")
		return
	}

	query := `
		SELECT m.id, m.nombre, m.created_at
		FROM marca m
		INNER JOIN marca_fabricante mf ON m.id = mf.marca_id
		WHERE mf.fabricante_id = ?
		ORDER BY m.nombre ASC
	`

	rows, err := database.ExecuteQuery(query, fabricanteID)
	if err != nil {
		respondWithError(w, http.StatusInternalServerError, "Error al obtener marcas")
		return
	}
	defer rows.Close()

	var marcas []models.Marca
	for rows.Next() {
		var marca models.Marca
		if err := rows.Scan(&marca.ID, &marca.Nombre, &marca.CreatedAt); err == nil {
			marcas = append(marcas, marca)
		}
	}

	respondWithJSON(w, http.StatusOK, marcas)
}

File: backend/routes/marcas.py
from flask import Blueprint, request, jsonify
from database import Database
from utils import MarcaValidator, FabricanteValidator, ValidationError, get_logger

marcas_bp = Blueprint('marcas', __name__)
db = Database()
logger = get_logger(__name__)


@marcas_bp.route('/api/marcas', methods=['GET'])
def get_marcas():
    """Obtiene todas las marcas con sus fabricantes asociados"""
    try:
        query = """
            SELECT m.id, m.nombre, 
                GROUP_CONCAT(f.id) as fabricante_ids,
                GROUP_CONCAT(f.nombre) as fabricante_nombres
            FROM marca m
            LEFT JOIN marca_fabricante mf ON m.id = mf.marca_id
            LEFT JOIN fabricantes f ON mf.fabricante_id = f.id
            GROUP BY m.id
        """
        
        marcas = db.execute_query(query, fetch_all=True)
        
        # Procesar fabricantes para cada marca
        for marca in marcas:
            if marca['fabricante_ids']:
                fabricante_ids = marca['fabricante_ids'].split(',')
                fabricante_nombres = marca['fabricante_nombres'].split(',')
                marca['fabricantes'] = [
                    {'id': int(id), 'nombre': nombre}
                    for id, nombre in zip(fabricante_ids, fabricante_nombres)
                ]
            else:
                marca['fabricantes'] = []
                
            # Eliminar campos innecesarios
            del marca['fabricante_ids']
            del marca['fabricante_nombres']
        
        return jsonify(marcas), 200
        
    except Exception as e:
        logger.error(f"Error obteniendo marcas: {e}", exc_info=True)
        return jsonify({"error": "Error al obtener marcas"}), 500


@marcas_bp.route('/api/marcas', methods=['POST'])
def add_marca():
    """Crea una nueva marca con fabricantes asociados"""
    try:
        data = request.get_json()
        
        if not data or 'nombre' not in data:
            return jsonify({"error": "El nombre de la marca es requerido"}), 400
        
        # Validar datos
        MarcaValidator.validate(data)
        
        # Insertar marca
        insert_query = """
            INSERT INTO marca (nombre)
            VALUES (%s)
        """
        marca_id = db.execute_insert(insert_query, (data['nombre'],))
        
        # Agregar asociaciones con fabricantes si se proporcionan
        if 'fabricanteIds' in data and data['fabricanteIds']:
            values = [(marca_id, fab_id) for fab_id in data['fabricanteIds']]
            
            fabricante_query = """
                INSERT INTO marca_fabricante (marca_id, fabricante_id)
                VALUES (%s, %s)
            """
            
            for marca_id_val, fab_id in values:
                db.execute_insert(fabricante_query, (marca_id_val, fab_id))
        
        return jsonify({
            "id": marca_id,
            "message": "Marca creada exitosamente"
        }), 201
        
    except ValidationError as e:
        logger.warning(f"Validación fallida al crear marca: {e}")
        return jsonify({"error": str(e)}), 400
    except Exception as e:
        logger.error(f"Error creando marca: {e}", exc_info=True)
        return jsonify({"error": "Error al crear marca"}), 500


@marcas_bp.route('/api/marcas/<int:marca_id>', methods=['PUT'])
def update_marca(marca_id):
    """Actualiza una marca existente"""
    try:
        data = request.get_json()
        
        if 'nombre' in data:
            update_query = """
                UPDATE marca 
                SET nombre = %s
                WHERE id = %s
            """
            db.execute_update(update_query, (data['nombre'], marca_id))
        
        if 'fabricanteIds' in data:
            # Eliminar asociaciones existentes
            delete_query = "DELETE FROM marca_fabricante WHERE marca_id = %s"
            db.execute_update(delete_query, (marca_id,))
            
            # Agregar nuevas asociaciones
            if data['fabricanteIds']:
                values = [(marca_id, fab_id) for fab_id in data['fabricanteIds']]
                
                insert_query = """
                    INSERT INTO marca_fabricante (marca_id, fabricante_id)
                    VALUES (%s, %s)
                """
                
                for marca_id_val, fab_id in values:
                    db.execute_insert(insert_query, (marca_id_val, fab_id))
        
        return jsonify({"message": "Marca actualizada exitosamente"}), 200
        
    except Exception as e:
        logger.error(f"Error actualizando marca {marca_id}: {e}", exc_info=True)
        return jsonify({"error": "Error al actualizar marca"}), 500


@marcas_bp.route('/api/fabricantes', methods=['GET'])
def get_fabricantes():
    """Obtiene todos los fabricantes"""
    try:
        query = "SELECT * FROM fabricantes ORDER BY nombre"
        fabricantes = db.execute_query(query, fetch_all=True)
        return jsonify(fabricantes), 200
        
    except Exception as e:
        logger.error(f"Error obteniendo fabricantes: {e}", exc_info=True)
        return jsonify({"error": "Error al obtener fabricantes"}), 500


@marcas_bp.route('/api/fabricantes', methods=['POST'])
def add_fabricante():
    """Crea un nuevo fabricante"""
    try:
        data = request.get_json()
        
        # Validar datos
        FabricanteValidator.validate(data)
        
        insert_query = """
            INSERT INTO fabricantes (nombre, pais)
            VALUES (%s, %s)
        """
        
        fabricante_id = db.execute_insert(
            insert_query,
            (data['nombre'], data.get('pais', ''))
        )
        
        return jsonify({
            "id": fabricante_id,
            "message": "Fabricante creado exitosamente"
        }), 201
        
    except ValidationError as e:
        logger.warning(f"Validación fallida al crear fabricante: {e}")
        return jsonify({"error": str(e)}), 400
    except Exception as e:
        logger.error(f"Error creando fabricante: {e}", exc_info=True)
        return jsonify({"error": "Error al crear fabricante"}), 500
